pkg/cache/v3/linear_test.go

```diff
@@ -414,49 +458,98 @@ func TestLinearVersionPrefix(t *testing.T) {
 	sub := subFromRequest(req)
 	_, err := c.CreateWatch(req, sub, w)
 	require.NoError(t, err)
-	verifyResponse(t, w, "instance1-0", 0)
+	resp := verifyResponse(t, w, "instance1-0", 0)
+	updateFromSotwResponse(resp, &sub, req)
 
 	require.NoError(t, c.UpdateResource("a", testResource("a")))
 
-	req.VersionInfo = "instance1-0"
 	_, err = c.CreateWatch(req, sub, w)
 	require.NoError(t, err)
-	verifyResponse(t, w, "instance1-1", 1)
+	resp = verifyResponse(t, w, "instance1-1", 1)
 
 	req.VersionInfo = "instance1-1"
+	sub.SetReturnedResources(resp.GetReturnedResources())
+	w = make(chan Response, 1)
 	_, err = c.CreateWatch(req, sub, w)
 	require.NoError(t, err)
 	mustBlock(t, w)
 	checkWatchCount(t, c, "a", 1)
 }
 
 func TestLinearDeletion(t *testing.T) {
-	c := NewLinearCache(testType, WithInitialResources(map[string]types.Resource{"a": testResource("a"), "b": testResource("b")}))
-	w := make(chan Response, 1)
-	req := &Request{ResourceNames: []string{"a"}, TypeUrl: testType, VersionInfo: "0"}
-	sub := subFromRequest(req)
-	_, err := c.CreateWatch(req, sub, w)
-	require.NoError(t, err)
-	mustBlock(t, w)
-	checkWatchCount(t, c, "a", 1)
+	t.Run("non full-state resource", func(t *testing.T) {
+		c := NewLinearCache(testType, WithInitialResources(map[string]types.Resource{"a": testResource("a"), "b": testResource("b")}))
+		w := make(chan Response, 1)
+		req := &Request{ResourceNames: []string{"a"}, TypeUrl: testType, VersionInfo: "0"}
+		sub := subFromRequest(req)
+		sub.SetReturnedResources(map[string]string{"a": "0"})
+		cancel, err := c.CreateWatch(req, sub, w)
+		require.NoError(t, err)
+		mustBlock(t, w)
+		checkWatchCount(t, c, "a", 1)
 
-	require.NoError(t, c.DeleteResource("a"))
-	verifyResponse(t, w, "1", 0)
-	checkWatchCount(t, c, "a", 0)
+		require.NoError(t, c.DeleteResource("a"))
+		// For non full-state type, we don't return on deletion
+		mustBlock(t, w)
+		// We currently do not clean watches for resources on deletion.
```

Comment:
Do you mean that you should? Is that why you left the next line commented out?

Reply:
Actually this comment is wrong, cleaned up